### PR TITLE
Fail subscirption updates by the customer on webhook

### PR DESCRIPTION
**Pull request by santhoshb-msft**
Result of the changes for below test cases

### 1. AcceptSubscriptionUpdates : True

**1.1 Change Plan from Azure portal**
image
image

**1.2 Change Quantity from Azure portal**
image
image

**1.3 Change Plan from Admin portal**
image
image

**1.4 Change Quantity from Admin portal**
image
image

**1.5 Change Plan from Customer portal**
image
image

**1.6 Change Quantity from Customer portal**
image
image

### 2. AcceptSubscriptionUpdates : False

**2.1 Change Plan from Azure portal**
image
image

**2.2 Change Quantity from Azure portal**
image
image

**2.3 Change Plan from Admin portal**
image
image

**2.4 Change Quantity from Admin portal**
image
image

**2.5 Change Plan from Customer portal**
image
image

**2.6 Change Quantity from Customer portal**
image
image

### CODE REVIEW

**Comment by santhoshb-msft** [comment]
#212

File: src/SaaS.SDK.CustomerProvisioning/WebHook/WebhookHandler.cs [hunk 1]
```diff
@@ -177,23 +195,41 @@ public async Task ChangePlanAsync(WebhookPayload payload)
...
-                    OldValue = oldValue.Quantity.ToString(),
+                    Attribute = Convert.ToString(SubscriptionLogAttributes.QuantityByWebhook),
+                    SubscriptionId = oldValue?.SubscribeId,
+                    OldValue = oldValue?.Quantity.ToString(),
                     CreateBy = null,
                     CreateDate = DateTime.Now,
                 };
-                this.subscriptionsLogRepository.Save(auditLog);
+
+                if (rejectSubscriptionUpdates)
```
Comment (code4clouds) [inline src/SaaS.SDK.CustomerProvisioning/WebHook/WebhookHandler.cs:212]: Add a small comment to describe this if (just to make it clear this a user configuration)

File: src/SaaS.SDK.CustomerProvisioning/WebHook/WebhookHandler.cs [hunk 2]
```diff
@@ -205,9 +241,46 @@ public async Task ChangeQuantityAsync(WebhookPayload payload)
...
+                SubscriptionAuditLogs auditLog = new SubscriptionAuditLogs()
+                {
+                    Attribute = Convert.ToString(SubscriptionLogAttributes.StatusByWebhook),
+                    SubscriptionId = oldValue?.SubscribeId,
+                    OldValue = Convert.ToString(oldValue?.SaasSubscriptionStatus),
+                    CreateBy = null,
+                    CreateDate = DateTime.Now,
+                };
+
+                if (rejectSubscriptionUpdates)
```
Comment (code4clouds) [inline src/SaaS.SDK.CustomerProvisioning/WebHook/WebhookHandler.cs:260]: Add a small comment to describe this if (just to make it clear this a user configuration)

File: src/SaaS.SDK.CustomerProvisioning/WebHook/WebhookHandler.cs [hunk 3]
```diff
@@ -216,9 +289,27 @@ public Task ReinstatedAsync(WebhookPayload payload)
         /// <param name="payload">The payload.</param>
         /// <returns> Exception.</returns>
         /// <exception cref="NotImplementedException"> Implemented Exception.</exception>
-        public Task SuspendedAsync(WebhookPayload payload)
+        public async Task SuspendedAsync(WebhookPayload payload)
         {
-            throw new NotImplementedException();
```
Comment (code4clouds) [inline src/SaaS.SDK.CustomerProvisioning/WebHook/WebhookHandler.cs:221]: Thanks for implementing this!

**Review by code4clouds — CHANGES_REQUESTED** [review]
Thanks for this submission... This is great.  Adding @dstarr  for a 2nd peer review as this is sizeable change on the webhook and also something he may want to know as part of the training.

**Comment by code4clouds** [comment]
@santhoshb-msft  Does the Admin webapp allow changes even when the RejectSubscriptionUpdates is True?

**Comment by santhoshb-msft** [comment]
> @santhoshb-msft Does the Admin webapp allow changes even when the RejectSubscriptionUpdates is True?

good q, Julio( @code4clouds), I will test that , it might not we may need to add some logic there to allow that. Let me check.

**Comment by code4clouds** [comment]
> > @santhoshb-msft Does the Admin webapp allow changes even when the RejectSubscriptionUpdates is True?
> 
> good q, Julio( @code4clouds), I will test that , it might not we may need to add some logic there to allow that. Let me check.

Any update on this?

**Comment by santhoshb-msft** [comment]
> > > @santhoshb-msft Does the Admin webapp allow changes even when the RejectSubscriptionUpdates is True?
> > 
> > 
> > good q, Julio( @code4clouds), I will test that , it might not we may need to add some logic there to allow that. Let me check.
> 
> Any update on this?

We used to have something in the webhook to recognize the action but currently we are not getting that in the webhook.  So even admin app changes wont be allowed.

**Comment by code4clouds** [comment]
@santhoshb-msft - We may have to document that if a plan change is required the publisher will have to:
1. disable the feature
2. restart admin webapp
3. make the plan change
then switch back on this feature and restart again.

**Comment by santhoshb-msft** [comment]
> @santhoshb-msft - We may have to document that if a plan change is required the publisher will have to:
> 
> 1. disable the feature
> 2. restart admin webapp
> 3. make the plan change
>    then switch back on this feature and restart again.

I have updated the messaging in the error itself. 
But sure, will add some documentation. There should no need to stop/start apps. Just updating the setting should have the feature working.

**Review by msalemcode — APPROVED** [review]

**Review by code4clouds — APPROVED** [review]

**Review by code4clouds — APPROVED** [review]

**Pull request merged**